src/Region.cpp

```diff
@@ -0,0 +1,808 @@
+          Halfspace* iter2 = dynamic_cast<Halfspace*>(*iter1);
+          if (iter2->getSurface()->getId() == surface->getId() && 
+              iter2->getHalfspace() == halfspace) {
+
+            delete iter2;
+            _nodes.erase(iter1);
+        }
+      }
+    }
+  }
+}
+
+
+/**
+ * @brief Return a vector of all of the Region's immediate nodes.
+ * @details This method will return a list of the Region's nodes
+ *          when called from Python.
+ * @returns a vector of the Region's immediate nodes
+ */
+std::vector<Region*> Region::getNodes() {
+  return _nodes;
+}
+
+
+/**
+ * @brief Return a vector of all of the Region's nodes.
+ * @details This method will return a list of the Region's nodes,
+ *          including nodes at lower levels.
+ * @returns a vector of the Region's nodes
+ */
+std::vector<Region*> Region::getAllNodes() {
+  std::vector<Region*> all_nodes;
+  std::vector<Region*> nodes;
+  std::vector<Region*>::iterator iter;
+  std::vector<Region*>::iterator sub_iter;
+   /* Recursively collect all nodes from this Regions nodes */
+  for (iter = _nodes.begin(); iter != _nodes.end(); iter++) {
+    all_nodes.push_back(*iter);
+    nodes = (*iter)->getAllNodes();
+    for (sub_iter = nodes.begin(); sub_iter != nodes.end(); sub_iter++)
+      all_nodes.push_back(*sub_iter);
+  }
+   return all_nodes;
+}
+
+
+/**
+ * @brief Extracts a map of all Halfspaces contained in the Region.
+ * @details This method recurses through all of the Region's nodes
+ *          and collects the Halfspaces into a map indexed by 
+ *          Surface ID. This method will return a dictionary of the
+ *          Halfspaces indexed by Surface ID called in Python.
+ * @returns a map of all Halfspaces in the Region
+ */
+std::map<int, Halfspace*> Region::getAllSurfaces() {
+
+  std::map<int, Halfspace*> all_surfaces;
+  std::map<int, Halfspace*> node_surfaces;
+  std::vector<Region*>::iterator iter;
+  /* Recursively collect all Halfspace's from this Regions nodes */
+  for (iter = _nodes.begin(); iter != _nodes.end(); iter++) {
+    node_surfaces = (*iter)->getAllSurfaces();
+    all_surfaces.insert(node_surfaces.begin(), node_surfaces.end());
+  }
+  return all_surfaces;
+}
+
+
+/**
+ * @brief Return the type of Region (ie, UNION, INTERSECTION, etc).
+ * @return the Region type
+ */
+regionType Region::getRegionType() {
+  return _region_type;
+}
+
+
+
+/**
+ * @brief Save the parent of the current node/Region.
+ * @param parent the node/Region that contains the current node/Region
+ */
+void Region::setParentRegion(Region* parent) {
+  _parent_region = parent;
+ }
+
+
+/**
+ * @brief Get the parent of the current node/Region.
+ * @return _parent_region the node/Region that contains the current node/Region
+ */
+Region* Region::getParentRegion() {
+   return _parent_region;
+}
+
+
+/**
+ * @brief Return the minimum reachable x-coordinate in the Region.
+ * @return the minimum x-coordinate
+ */
+double Region::getMinX() {
```

Comment:
Supposedly, currently it might only be right for all intersection case.

Reply:
double Region::getMinX() {

  /* Set a default min-x */
  double min_x;
  if(_region_type == INTERSECTION)
    min_x = -std::numeric_limits<double>::infinity();
  else if(_region_type == Union)
    min_x = std::numeric_limits<double>::infinity();

  /* Loop over all nodes in the Region */
  std::vector<Region*>::iterator iter;
  for (iter = _nodes.begin(); iter != _nodes.end(); ++iter) {
    if(_region_type == INTERSECTION)
      min_x = std::max(min_x, (*iter)->getMinX());
    else if(_region_type == Union)
      min_x = std::min(min_x, (*iter)->getMinX());
  }
  return min_x;
}
consider this ?

Reply:
This looks good. I'll look into it.

Reply:
This will fail if we don't include all Halfspaces at each node. 
For a sector for example, if we don't have the the Z-cylinder from the parent cell this will break. 

I'll add something for this case